### DISCUSSION TITLE
How to resolve `could not infer the batch_size from the batch`

### DISCUSSION THREAD

**Discussion by Zhengxinyang** [Lightning-AI/pytorch-lightning#13683] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I encountered the following error report：

pytorch_lightning.utilities.exceptions.MisconfigurationException: We could not infer the batch_size from the batch. Either simplify its structure or provide the batch_size as `self.log(..., batch_size=batch_size)`.

How can I fix it without changing the data loader?

**Comment by akihironitta** [comment]
> How can I fix it without changing the data loader?

@Zhengxinyang You can fix the error by providing `the batch_size as self.log(..., batch_size=batch_size)` as the error suggests. (Or, are you asking something different?)

**Reply by Zhengxinyang** [reply]
Thanks. I have tried it but it doesn't work.  Should  I use `self.log(..., batch_size=batch_size)` in the train_dataloader or training_step？🧐

**Reply by akihironitta** [reply]
I think having it in your `training_step` or wherever the exception is raised within should fix the error.

FYI, here's the reason why you see the exception:

> While logging tensor metrics with on_epoch=True inside step-level hooks and using mean-reduction (default) to accumulate the metrics across the current epoch, Lightning tries to extract the batch size from the current batch. If multiple possible batch sizes are found, a warning is logged and if it fails to extract the batch size from the current batch, which is possible if the batch is a custom structure/collection, then an error is raised. To avoid this, you can specify the batch_size inside the self.log(... batch_size=batch_size) call.

https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#id3

**Reply by Zhengxinyang** [reply]
Thanks. In fact, I found this error triggered by `Trainer(... , track_grad_norm=2)`, how can I fix it?

**Reply by akihironitta** [reply]
@Zhengxinyang `track_grad_norm` shouldn't impact it at all AFAIK. I've tried to reproduce it with PL master, however, I cannot reproduce the behaviour. https://github.com/akihironitta/gist/blob/repro/13683-track-grad-norm-batch-size/pl_boring_model/main.py

Could you provide the env detail and script to reproduce?

**Reply by rubencart** [reply]
Running into the same error, also only when `track_grad_norm=2`. I think it is because tracking the gradient norm makes a call to `log_dict` without `batch_size=...`, see part of the stacktrace below.
```
  File ".../miniconda3/envs/satenv4/lib/python3.10/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 85, in _after_closure
    self._track_grad_norm(trainer)
  File ".../miniconda3/envs/satenv4/lib/python3.10/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 135, in _track_grad_norm
    trainer.lightning_module.log_grad_norm(grad_norm_dict)
  File ".../miniconda3/envs/satenv4/lib/python3.10/site-packages/pytorch_lightning/core/module.py", line 565, in log_grad_norm
    self.log_dict(grad_norm_dict, on_step=True, on_epoch=True, prog_bar=False, logger=True)
```

**Reply by janfer95** [reply]
Are there any news regarding this? I run into the same error when using `track_grad_norm=1` in a pytorch_geometric  / pytorch_lightning code, even though I do pass `self.log(..., batch_size=batch_size` in `training_step`. 

Is there maybe a way to pass it directly to `track_grad_norm`?

**Reply by So-cean** [reply]
same problem, when using torch_geometric. it seems that lightning treated pyg data's num_nodes as batch_size
